apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx

```diff
@@ -0,0 +1,203 @@
+import React from 'react';
+import Radium from 'radium';
+import PropTypes from 'prop-types';
+import {connect} from 'react-redux';
+import _ from 'lodash';
+import BaseDialog from '@cdo/apps/templates/BaseDialog.jsx';
+import DropdownField from './DropdownField';
+import * as dataStyles from '../dataStyles';
+import * as rowStyle from '@cdo/apps/applab/designElements/rowStyle';
+import DataVisualizer from './DataVisualizer';
+
+const INITIAL_STATE = {
+  isVisualizerOpen: false,
+  chartTitle: '',
+  chartType: '',
+  bucketSize: '',
+  column1: '',
+  column2: '',
+  parsedRecords: [],
+  numericColumns: []
+};
+
+class VisualizerModal extends React.Component {
+  static propTypes = {
+    // from redux state
+    tableColumns: PropTypes.arrayOf(PropTypes.string).isRequired,
+    tableName: PropTypes.string.isRequired,
+    // "if all of the keys are integers, and more than half of the keys between 0 and
+    // the maximum key in the object have non-empty values, then Firebase will render
+    // it as an array."
+    // https://firebase.googleblog.com/2014/04/best-practices-arrays-in-firebase.html
+    tableRecords: PropTypes.oneOfType([PropTypes.object, PropTypes.array])
+      .isRequired
+  };
+
+  state = {...INITIAL_STATE};
+
+  handleOpen = () => {
+    this.setState({isVisualizerOpen: true});
+  };
+
+  handleClose = () => {
+    this.setState({...INITIAL_STATE});
+  };
+
+  parseRecords = () => {
+    if (Object.keys(this.props.tableRecords).length === 0) {
+      return [];
+    } else {
+      let parsedRecords = [];
+      this.props.tableRecords.forEach(record => {
+        if (record) {
+          parsedRecords.push(JSON.parse(record));
+        }
+      });
+      return parsedRecords;
+    }
+  };
+
+  isCellEmpty = value => {
+    return value === undefined || value === '' || value === null;
+  };
+
+  findNumericColumns = records => {
+    return this.props.tableColumns.filter(
+      column =>
+        records.filter(
+          record =>
+            this.isCellEmpty(record[column]) ||
+            typeof record[column] === 'number'
+        ).length === records.length
+    );
+  };
+
+  componentDidUpdate(previousProps, prevState) {
+    const visualizerJustOpened =
+      !prevState.isVisualizerOpen && this.state.isVisualizerOpen;
+
+    const recordsChangedWithVisualizerOpen =
+      this.props.tableRecords !== previousProps.tableRecords &&
+      this.state.isVisualizerOpen;
+
+    if (visualizerJustOpened || recordsChangedWithVisualizerOpen) {
+      let parsedRecords = this.parseRecords();
+      let numericColumns = this.findNumericColumns(parsedRecords);
+      this.setState({
+        parsedRecords: parsedRecords,
+        numericColumns: numericColumns
+      });
```

Comment:
A possible optimization and simplification:  As written, I think you always trigger two renders back-to-back when the visualizer opens or records change. A better way to avoid redoing expensive computation is with a [memoized function](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#what-about-memoization).  It will take care of keeping a copy of the latest results around and serving it up as needed, and deciding when they need to be recomputed. Just [don't use lodash's `_.memoize` for this](https://dev.to/nioufe/you-should-not-use-lodash-for-memoization-3441).